### ISSUE TITLE
any suggestion on how to get rid of the checkered effect 

### ISSUE THREAD

**Issue by AllAwake** [junyanz/pytorch-CycleGAN-and-pix2pix#190]
i'm training CycleGAN on my food photos and my sketches - the results are fascinating esp colors and shapes  - if only i knew how to get rid of that checkered effect...
![38744547705_046ce1d663_o](https://user-images.githubusercontent.com/4332614/35075502-55e28c62-fbc1-11e7-9a7d-0ab82d7c1227.jpg)
![39619927772_e0c97ed1b0_o](https://user-images.githubusercontent.com/4332614/35075503-55ef78aa-fbc1-11e7-868a-5cb094ac5166.jpg)

**Comment by phillipi** [comment]
Very cool! Care to share what the inputs look like too?

This [distill paper](https://distill.pub/2016/deconv-checkerboard/) talks about one of the causes of the checkerboard artifacts. You can fix that issue by switching from "deconvolution" to nearest-neighbor upsampling followed by regular convolution. I think @SsnL may have implemented this at some point.

We've also noticed that sometimes the checkboard artifacts go away if you simply train long enough. Maybe try training a bit longer.

Another cause of repetitive artifacts can be that the discriminator's receptive field is too small. For some discussion on this, please see Section 4.4 and Figure 6 of the [pix2pix paper](https://arxiv.org/pdf/1611.07004.pdf). The issue is that if the discriminator looks at too myopic a region, it won't notice that textures are repeating. I think this is probably not the case in your results, but it's something to keep in mind.

**Comment by SsnL** [comment]
@AllAwake Cool results! 

Here is the implementation of resize-conv I used. It remove the checkerboard artifacts during early training. You may find it useful.
```
                          nn.Upsample(scale_factor = 2, mode='bilinear'),
                          nn.ReflectionPad2d(1),
                          nn.Conv2d(ngf * mult, int(ngf * mult / 2),
                                             kernel_size=3, stride=1, padding=0),
```
It should replace the `ConvTranspose2d` in `ResnetGenerator`.

**Comment by codaibk** [comment]
Hi @SsnL and @phillipi 
I currently research about `lower-to-high resolution `image. My `Generato`r and `Discriminator` artichecture are same as your cycleGan. I followed your guide to replace the `ConvTranspose2d` but it seems the checkerboard artifacts still appear in my result. Follow the paper [distill](https://distill.pub/2016/deconv-checkerboard/), they mentioned about `resize the image` (using nearest-neighbor interpolation or bilinear interpolation) and also changing something in `Discriminato`r. Could you please tell me how we implement it to remove the checkerboard artifacts?. This is my result
Input 
![116_real_a](https://user-images.githubusercontent.com/28720676/48473070-a63fad80-e832-11e8-977c-4c98fbf8aa12.png)
result
![116_fake_b2](https://user-images.githubusercontent.com/28720676/48473092-afc91580-e832-11e8-9e46-8aac541d2c46.png)

**Comment by junyanz** [comment]
I haven't tried the distill tricks by myself. For discriminators, they mentioned that you can replace stride 2 [conv](https://github.com/junyanz/pytorch-CycleGAN-and-pix2pix/blob/master/models/networks.py#L325) with a regular 3x3 conv.

**Comment by codaibk** [comment]
> I haven't tried the distill tricks by myself. For discriminators, they mentioned that you can replace stride 2 [conv](https://github.com/junyanz/pytorch-CycleGAN-and-pix2pix/blob/master/models/networks.py#L325) with a regular 3x3 conv.

Does Regular 3x3 conv mean that we just need to change stride =1 in this case ?

**Comment by jacky841102** [comment]
> @AllAwake Cool results!
> 
> Here is the implementation of resize-conv I used. It remove the checkerboard artifacts during early training. You may find it useful.
> 
> ```
>                           nn.Upsample(scale_factor = 2, mode='bilinear'),
>                           nn.ReflectionPad2d(1),
>                           nn.Conv2d(ngf * mult, int(ngf * mult / 2),
>                                              kernel_size=3, stride=1, padding=0),
> ```
> It should replace the `ConvTranspose2d` in `ResnetGenerator`.

Hi @SsnL 
may I ask why you choose bilinear upsampling instead of nearest-neighbor one?  The [distill paper](https://distill.pub/2016/deconv-checkerboard/) pointed out the result of nearest-neighbor interpolation should be better.

**Comment by SsnL** [comment]
@jacky841102 You could try nearest neighbor. I think that it didn’t make much difference in the dataset I tried.

**Comment by mrgloom** [comment]
I wonder why pix2pixHD is using `nn.ConvTranspose2d` anyway?
https://github.com/NVIDIA/pix2pixHD/issues/149

Same with StyleGAN
https://github.com/NVlabs/stylegan/blob/f3a044621e2ab802d40940c16cc86042ae87e100/training/networks_stylegan.py#L174

and ProGAN
https://github.com/NVlabs/stylegan/blob/f3a044621e2ab802d40940c16cc86042ae87e100/training/networks_progan.py#L89

**Comment by mrgloom** [comment]
Also according to this http://warmspringwinds.github.io/tensorflow/tf-slim/2016/11/22/upsampling-and-image-segmentation-with-tensorflow-and-tf-slim/
transposed convolution can be initialized with bilinear filter.

**Comment by yanfei-zhang-95** [comment]
> @AllAwake Cool results!
> 
> Here is the implementation of resize-conv I used. It remove the checkerboard artifacts during early training. You may find it useful.
> 
> ```
>                           nn.Upsample(scale_factor = 2, mode='bilinear'),
>                           nn.ReflectionPad2d(1),
>                           nn.Conv2d(ngf * mult, int(ngf * mult / 2),
>                                              kernel_size=3, stride=1, padding=0),
> ```
> 
> It should replace the `ConvTranspose2d` in `ResnetGenerator`.

This method works pretty well, but it would also cause unstable training, therefore I applied an exponential learning rate decay and small initial learning rate to ensure that it works well enough

**Comment by AlexTS1980** [comment]
@SsnL Thanks for the solution. Did anyone successfully implement it in Generators that monotonically upsample the input? In this example, for example, if you add a padding, the output will be **exactly** twice the size of the input;
```
nn.Upsample(scale_factor = 2, mode='bilinear'),
                          nn.ReflectionPad2d(1),
                          nn.Conv2d(ngf * mult, int(ngf * mult / 2),
                                             kernel_size=3, stride=1, padding=1),
```

**Comment by AlexTS1980** [comment]
> @AllAwake Cool results!
> 
> Here is the implementation of resize-conv I used. It remove the checkerboard artifacts during early training. You may find it useful.
> 
> ```
>                           nn.Upsample(scale_factor = 2, mode='bilinear'),
>                           nn.ReflectionPad2d(1),
>                           nn.Conv2d(ngf * mult, int(ngf * mult / 2),
>                                              kernel_size=3, stride=1, padding=0),
> ```
> 
> It should replace the `ConvTranspose2d` in `ResnetGenerator`.

I like how it totally gets rid of the checkerboard artefacts, but it comes at a cost of actually learning useful features. I'm trying to balance these effects.
